### ISSUE TITLE
`new_without_default_derive` triggers on types which cannot `#[derive(Default)]`

### ISSUE THREAD

**Issue by sgrif** [rust-lang/rust-clippy#1890]
The lint seems to trigger on *any* function with the signature `fn new() -> Self`. `#[derive(Default)]` requires a default bound on all type parameters, even if those types do not appear in the actual type body.

**Comment by oli-obk** [comment]
The idea of that lint is not to necessarily use derive, but to implement Default by forwarding to new

**Comment by sgrif** [comment]
If that's the case, it should probably not include `derive` in the name or the help text

**Comment by oli-obk** [comment]
Oh sorry. I confused the two `new_without_default*` lints. That makes total sense.

**Comment by oli-obk** [comment]
Can you give an example struct where the lint triggers but shouldn't?

**Comment by sgrif** [comment]
```rust
use std::collections::HashMap;

trait SomeTrait {
    type SomeType;
}

#[derive(Hash, PartialEq, Eq)]
struct Key<T: SomeTrait> {
    foo: T::SomeType,
}

struct Cache<T: SomeTrait> {
    foo: HashMap<Key<T>, ()>,
}

impl<T> Cache<T>
where
    T: SomeTrait,
    T::SomeType: Hash + Eq,
{
    pub fn new() -> Self {
        Cache {
            foo: HashMap::new(),
        }
    }
}
```

**Comment by oli-obk** [comment]
Unfortunately I cannot reproduce: http://play.integer32.com/?gist=56b78c856f90a1f8e0bb9e8ee50a392a&version=stable or http://play.integer32.com/?gist=8179a9caaadbd71712a54d804fea13e5&version=stable (your code as it is doesn't compile, so I had to guess)

**Comment by sgrif** [comment]
Sorry about that, this is the code which compiles:

```rust
use std::collections::HashMap;
use std::hash::Hash;

trait SomeTrait {
    type SomeType;
}

#[derive(Hash, PartialEq, Eq)]
struct Key<T: SomeTrait> {
    foo: T::SomeType,
}

struct Cache<T: SomeTrait> {
    foo: HashMap<Key<T>, ()>,
}

impl<T> Cache<T>
where
    T: SomeTrait,
    Key<T>: Hash + Eq,
{
    pub fn new() -> Self {
        Cache {
            foo: HashMap::new(),
        }
    }
}

fn main() {
}
```

**Comment by oli-obk** [comment]
But it doesn't trigger the lint. At least not on play.integer32.com

**Comment by sgrif** [comment]
The actual type which 100% triggers the lint and cannot derive Default which I was attempting to simplify is [here](https://github.com/diesel-rs/diesel/blob/596d3ad9d0c73cff265ecd7887dfd6cf44412d15/diesel/src/connection/statement_cache.rs#L104-L108)

**Comment by oli-obk** [comment]
I consider this a bug in derives, not in clippy: https://github.com/rust-lang/rust/issues/9607

minimal repro:

```rust
use std::hash::Hash;
use std::collections::HashMap;

pub struct Foo<T> {
    bar: HashMap<T, usize>,
}

impl<T: Hash + Eq> Foo<T> {
    pub fn new() -> Self {
        Foo {
            bar: HashMap::default(),
        }
    }
}
```

**Comment by radix** [comment]
I'm not sure if this should be a separate issue or not, but I feel like this code should *not* trigger this lint:


```rust
#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
pub struct MapID(Uuid);
impl MapID {
  pub fn new() -> MapID {
    MapID(Uuid::new_v4())
  }
  pub fn to_string(&self) -> String {
    self.0.hyphenated().to_string()
  }
}
```

IMO it should only trigger it when the contents of the `new` method would be equivalent to using a derived Default.

**Comment by hellow554** [comment]
I think my case is related to this bug. The following will produce the `new_without_default_derive` error, which it shouldn't.

```rust
#[derive(Debug)]
pub struct Foo<'a, T> {
    bar: Vec<&'a T>,
}

impl<'a, T> Foo<'a, T> {
    pub fn new() -> Self {
        Foo { bar: Vec::new() }
    }
}

fn main() {}
```

```none
warning: you should consider deriving a `Default` implementation for `Foo<'a, T>`
 --> src/main.rs:7:5
  |
7 | /     pub fn new() -> Self {
8 | |         Foo { bar: Vec::new() }
9 | |     }
  | |_____^
  |
  = note: #[warn(clippy::new_without_default_derive)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#new_without_default_derive
help: try this
  |
2 | #[derive(Default)]
  |
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=75e571f75ace01ffc171aa6f2c68864a))

**Comment by rail-rain** [comment]
For anyone who happened to be here: `new_without_dafault_derive` was merged into `new_without_default` in #3558, and then it stopped suggesting deriving `Default` in #5616. Therefore, unless someone re-implemented the feature with [the enhancement suggested by flip1995](https://github.com/rust-lang/rust-clippy/issues/3697#issuecomment-458642843), there's nothing to be done for this issue.

By the way, the repro by oli-obk can derive `Default` now due to https://github.com/rust-lang/rust/pull/67642; which is unrelated to the problem.
